### PR TITLE
chore(RHTAPWATCH-592): Create a README in o11y/exporter/dsexporter/ and add blurb about exporters in the root README

### PR DESCRIPTION
**Pull request by Kousalya1998**
- Add README.md under o11y/exporter/dsexporter/
- Add blub about exporters in the root README.md

### CODE REVIEW

**Comment by yftacherzog** [comment]
@Kousalya1998 can you break all lines at ~90 chars?

File: README.md [hunk 1]
```diff
@@ -99,8 +100,35 @@ to RHOBS. If additional metrics or labels are needed, add them by following the
...
+### Availability Exporter Example
+The o11y team provides an example availability exporter that can be used as reference,
+especially in the case in which the exporter is external to the code it's monitoring.
+
+- [Exporter code](https://github.com/redhat-appstudio/o11y/tree/main/exporters/dsexporter)
+- [Exporter and service Monitor Kubernetes Resources](https://github.com/redhat-appstudio/o11y/tree/main/config/exporters/monitoring/grafana/base)
+
+### References
+
+- [Availability exporters deatiled documentation](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md?ref_type=heads)
```
Comment (eisraeli) [inline README.md:128]:
```suggestion
- [Availability exporters detailed documentation](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md?ref_type=heads)
```

File: README.md [hunk 2]
```diff
@@ -99,8 +100,35 @@ to RHOBS. If additional metrics or labels are needed, add them by following the
...
+In order to be able to evaluate the overall availability of the Konflux ecosystem, we
+need to be able to establish the availability of each of its components.
+
+By leveraging the existing [Konflux monitoring stack](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/monitoring.md) based on Prometheus, we create Prometheus exporters that 
+generate metrics that are scraped by the User Workload Monitoring Prometheus instance 
+and remote-written to RHOBS.
+
+### Prometheus Exporters
+
+Prometheus exporters expose metrics from a system or application in a format that can
```
Comment (eisraeli) [inline README.md:114]:
```suggestion
Prometheus exporters expose metrics from a system or an application in a format that can 
```

File: exporters/dsexporter/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Purpose
+
+In the context of the Konflux ecosystem, the `dsexporter` serves as an Availability Exporter, contributing to the
+evaluation of the availability of each component within the system. By exposing the `request_count` metric, it allows
+for monitoring and analyzing the workload and performance of the associated service.
+
+## Prometheus Exporters
+
+A **Prometheus exporter** exposes metrics from a system or application in a format that Prometheus can ingest and monitor.
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:17]:
```suggestion
A [Prometheus exporter](https://prometheus.io/docs/instrumenting/exporters/#exporters-and-integrations) exposes metrics from a system or application in a format that Prometheus can ingest and monitor.
```

File: exporters/dsexporter/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+The `dsexporter` follows the Prometheus exposition format, providing a web endpoint that outputs metrics in a well-known format.
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:18]: There are actually two exposition formats: 
1. Text-based format.
2. OpenMetrics Text Format.

It's be more specific here.

File: exporters/dsexporter/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+## Getting Started
+
+To use the `dsexporter`, follow these steps:
+
+1. Import the necessary packages in your code.
+2. Start the HTTP server and the exporter will be accessible at the respected endpoint.
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:25]: While the concept of `dsexporter`s was explained, I feel as this section in lacking details.

File: exporters/dsexporter/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+## Overview
+
+The `dsexporter` is a specialized Prometheus exporter designed to provide availability metrics for the components
+within the Konflux ecosystem. It generates a custom metric (`request_count`) related to the number of requests handled
+by an HTTP handler, enabling the evaluation of overall availability.
+
+## Purpose
+
+In the context of the Konflux ecosystem, the `dsexporter` serves as an Availability Exporter, contributing to the
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:11]:
```suggestion
In the context of the Konflux ecosystem, the `dsexporter` serves as an [Availability Exporter](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md), contributing to the
```

**Comment by eisraeli** [comment]
IMO, the changes in the root README.md are redundant as the information already exists [here](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md)
I think we should just reference to it.

File: exporters/dsexporter/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+evaluation of the availability of each component within the system. By exposing the `request_count` metric, it allows
+for monitoring and analyzing the workload and performance of the associated service.
+
+## Prometheus Exporters
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:15]: This information already exists here:
https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md?ref_type=heads#prometheus-exporters

File: exporters/dsexporter/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# dsexporter
+
+## Overview
+
+The `dsexporter` is a specialized Prometheus exporter designed to provide availability metrics for the components
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:5]: Maybe you can explain here what `ds` stands for ?

File: exporters/dsexporter/README.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+within the Konflux ecosystem. It generates a custom metric (`request_count`) related to the number of requests handled
```
Comment (hmariset) [inline exporters/dsexporter/README.md:6]: request_count is not relevant anymore since we are updating the code with actual metrics.

**Comment by yftacherzog** [comment]
> > @Kousalya1998 can you break all lines at ~90 chars?
> 
> Updated

Most lines are still very long.

File: exporters/dsexporter/README.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+This `dummy service exporter file (dsexporter.go)` exposes a single metric named "request_count" that counts the number of requests handled by the service. The exporter implements a custom Prometheus collector named `CustomCollector`, which implements the `prometheus.Collector` interface. The CustomCollector struct has a requestCounter field to track the number of requests.It implements the `Describe` method to provide descriptions of the metrics to Prometheus and `Collect` method to send the current values of the metrics to Prometheus.
+
+The `test case file (main_test.go)` tests the functionality of the CustomCollector. It verifies that the `request_count` metric initially has a value of 0 and then increments to 1 after simulating a request.
+
+This code provides a simple example of how to create a custom Prometheus exporter and test its functionality in Go. It is useful for monitoring the performance and behavior of the overall availability of the Konflux ecosystem.
+
+The o11y team provides an example availability exporter that can be used as reference below:
+
+* [Exporterd code](https://github.com/redhat-appstudio/o11y/tree/main/exporters/dsexporter)
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:23]:
```suggestion
* [Exporterr code](https://github.com/redhat-appstudio/o11y/tree/main/exporters/dsexporter)
```

File: exporters/dsexporter/README.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+* [Epxorter and Service Monitor Kubernetes Resources](https://github.com/redhat-appstudio/o11y/tree/main/config/exporters/monitoring/grafana/base)
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:24]:
```suggestion
* [Exporter and Service Monitor Kubernetes Resources](https://github.com/redhat-appstudio/o11y/tree/main/config/exporters/monitoring/grafana/base)
```

File: exporters/dsexporter/README.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+another system/component.
+
+## Availability Exporter
+
+In the context of the Konflux ecosystem, the `dsexporter` serves as an [Availability Exporter](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md), contributing to the evaluation of the availability of each component within the system. By exposing this metric, it allows for monitoring and analyzing the workload
+and performance of the associated service.
+
+## dsexporter
+
+This `dummy service exporter file (dsexporter.go)` exposes a single metric named "request_count" that counts the number of requests handled by the service. The exporter implements a custom Prometheus collector named `CustomCollector`, which implements the `prometheus.Collector` interface. The CustomCollector struct has a requestCounter field to track the number of requests.It implements the `Describe` method to provide descriptions of the metrics to Prometheus and `Collect` method to send the current values of the metrics to Prometheus.
```
Comment (eisraeli) [inline exporters/dsexporter/README.md:15]:
```suggestion
This `dummy service exporter file (dsexporter.go)` exposes a single metric named "request_count" that counts the number of requests handled by the service. The exporter implements a custom Prometheus collector named `CustomCollector`, which implements the `prometheus.Collector` interface. The CustomCollector struct has a requestCounter field to track the number of requests. It implements the `Describe` method to provide descriptions of the metrics to Prometheus and `Collect` method to send the current values of the metrics to Prometheus.
```

File: exporters/dsexporter/README.md (hunk 5 above)
Comment (eisraeli) [reply exporters/dsexporter/README.md]: Do you think this previous Getting Started paragraph is redundant now ? As I see it was completely removed.

File: exporters/dsexporter/README.md (hunk 7 above)
Comment (Kousalya1998) [reply exporters/dsexporter/README.md]: updated it now

File: exporters/dsexporter/README.md (hunk 5 above)
Comment (Kousalya1998) [reply exporters/dsexporter/README.md]: Keeping it brief and removing details was my thought.

**Comment by Kousalya1998** [comment]
> > > @Kousalya1998 can you break all lines at ~90 chars?
> > 
> > 
> > Updated
> 
> Most lines are still very long.

I updated the README now.

File: exporters/dsexporter/README.md (hunk 5 above)
Comment (eisraeli) [reply exporters/dsexporter/README.md]: Keep it brief is fine, but can this information be found anywhere else? 
Do you think this information is important to someone who will want to use dsexporter?  - if not, then I'm fine with this change.

File: exporters/dsexporter/README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+This `dummy service exporter file (dsexporter.go)` exposes a single metric named 
+"request_count" that counts the number of requests handled by the service. 
+The exporter implements a custom Prometheus collector named `CustomCollector`, 
+which implements the `prometheus.Collector` interface. The CustomCollector struct
+has a requestCounter field to track the number of requests. It implements the 
+`Describe` method to provide descriptions of the metrics to Prometheus and `Collect`
+method to send the current values of the metrics to Prometheus.
+
+The `test case file (main_test.go)` tests the functionality of the CustomCollector.
```
Comment (hmariset) [inline exporters/dsexporter/README.md:25]: Test file is named as `(dsexporter_test.go)`

**Review by hmariset — CHANGES_REQUESTED** [review]

**Comment by Kousalya1998** [comment]
> > > @Kousalya1998 can you break all lines at ~90 chars?
> > 
> > 
> > Updated
> 
> Most lines are still very long.

I don't know why for some reason my vscode shows like I have already indented them. I referred to the GitHubs README and updated the indentation again.

**Review by hmariset — APPROVED** [review]
lgtm

File: exporters/dsexporter/README.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+In the context of the Konflux ecosystem, the `dsexporter` serves as an 
+[Availability Exporter](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md),
+contributing to the evaluation of the availability of each component within the system.
+By exposing this metric, it allows for monitoring and analyzing the 
+workload and performance of the associated service.
+
+## dsexporter
+
+This `dummy service exporter file (dsexporter.go)` exposes a single metric named
```
Comment (amisstea) [inline exporters/dsexporter/README.md:18]: `dsexporter` stands for "datasource exporter", not "dummy service".

File: exporters/dsexporter/README.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+In the context of the Konflux ecosystem, the `dsexporter` serves as an 
+[Availability Exporter](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md),
+contributing to the evaluation of the availability of each component within the system.
+By exposing this metric, it allows for monitoring and analyzing the 
+workload and performance of the associated service.
+
+## dsexporter
+
+This `dummy service exporter file (dsexporter.go)` exposes a single metric named 
+"request_count" that counts the number of requests handled by the service.
```
Comment (amisstea) [inline exporters/dsexporter/README.md:19]: Take a look at the recent changes in #174. The metric is now `konflux_up`. Please update this description.

**Review by amisstea — CHANGES_REQUESTED** [review]

File: exporters/dsexporter/README.md (hunk 15 above)
Comment (Kousalya1998) [reply exporters/dsexporter/README.md]: @amisstea Updated. Need your reviews.

File: exporters/dsexporter/README.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+In the context of the Konflux ecosystem, the `dsexporter` serves as an 
+[Availability Exporter](https://gitlab.cee.redhat.com/konflux/docs/documentation/-/blob/main/o11y/monitoring/availability_exporters.md),
+contributing to the evaluation of the availability of each component within the system.
+By exposing this metric, it allows for monitoring and analyzing the 
+workload and performance of the associated service.
+
+## dsexporter
+
+This `datasource exporter file (dsexporter.go)` exposes a single metric named 
+"konflux_up" that counts the number of requests handled by the service.
```
Comment (amisstea) [inline exporters/dsexporter/README.md:19]: `konflux_up` doesn't count requests. Please update this description.

File: exporters/dsexporter/README.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+"konflux_up" that counts the number of requests handled by the service. 
+The exporter implements a custom Prometheus collector named `CustomCollector`, 
+which implements the `prometheus.Collector` interface. The CustomCollector struct
+has a requestCounter field to track the number of requests. It implements the 
+`Describe` method to provide descriptions of the metrics to Prometheus and `Collect`
+method to send the current values of the metrics to Prometheus.
+
+The `test case file (dsexporter_test.go)` tests the functionality of the CustomCollector. 
+It verifies that the `konflux_up` metric initially has a value of 0 and then 
+increments to 1 after simulating a request.
```
Comment (amisstea) [inline exporters/dsexporter/README.md:28]: This explanation also seems to be based on the old counter metric. Please update it.

Comment (Kousalya1998) [reply exporters/dsexporter/README.md:28]: Sure, I will check with Homaja again an update everything all at once.

Comment (Kousalya1998) [reply exporters/dsexporter/README.md:28]: @amisstea I wrote this README keeping in mind the current exporters/dsexporter/dsexporter.go in the upstream. If I have to write it for the latest dsexporter.go code I think I should wait on this PR for all the changes to be updated on the upstream.

Comment (Kousalya1998) [reply exporters/dsexporter/README.md:28]: The current dsexporter.go code has a request_count metric in the upstream.

Comment (amisstea) [reply exporters/dsexporter/README.md:28]: #174 has now been merged.

**Comment by Kousalya1998** [comment]
@amisstea @hmariset PTAL at the PR. Updated the README.

**Pull request merged**